src/combinatorics/derangements/basic.lean

```diff
@@ -0,0 +1,190 @@
+  ... ≃ {f : perm α // fixed_points f = {a}} ⊕ {f : perm α // fixed_points f = ∅}
+      : begin
+        refine equiv.sum_congr (subtype_equiv_right $ λ f, _) (subtype_equiv_right $ λ f, _),
+        { rw [set.eq_singleton_iff_unique_mem, and_comm],
+          refl },
+        { rw set.eq_empty_iff_forall_not_mem,
+          refine ⟨λ h x hx, h.2 (h.1 hx ▸ hx), λ h, ⟨λ x hx, (h _ hx).elim, h _⟩⟩ }
+      end
+  ... ≃ (derangements _) ⊕ derangements α
+      : begin
+        refine equiv.sum_congr ((derangements.subtype_equiv _).trans $ subtype_equiv_right $ λ x,
+          _).symm (subtype_equiv_right $ λ f, mem_derangements_iff_fixed_points_eq_empty.symm),
+        rw [eq_comm, set.ext_iff],
+        simp_rw [set.mem_compl_iff, not_not],
+      end
+
+namespace equiv
+variables [decidable_eq α]
+
+/-- The set of permutations `f` such that the preimage of `(a, f)` under
+    `equiv.perm.decompose_option` is a derangement. -/
+def remove_none.fiber (a : option α) : set (perm α) :=
+  {f : perm α | (a, f) ∈ equiv.perm.decompose_option '' derangements (option α)}
+
+lemma remove_none.mem_fiber (a : option α) (f : perm α) :
+  f ∈ remove_none.fiber a ↔
+  ∃ F : perm (option α), F ∈ derangements (option α) ∧ F none = a ∧ remove_none F = f :=
+by simp [remove_none.fiber, derangements]
+
+lemma remove_none.fiber_none : remove_none.fiber (@none α) = ∅ :=
+begin
+  rw set.eq_empty_iff_forall_not_mem,
+  intros f hyp,
+  rw remove_none.mem_fiber at hyp,
+  rcases hyp with ⟨F, F_derangement, F_none, _⟩,
+  exact F_derangement none F_none
+end
+
+/-- For any `a : α`, the fiber over `some a` is the set of permutations
+    where `a` is the only possible fixed point. -/
+lemma remove_none.fiber_some (a : α) :
+  (remove_none.fiber (some a)) = {f : perm α | fixed_points f ⊆ {a}} :=
+begin
+  ext f,
+  split,
+  { rw remove_none.mem_fiber,
+    rintro ⟨F, F_derangement, F_none, rfl⟩ x x_fixed,
+    rw mem_fixed_points_iff at x_fixed,
+    apply_fun some at x_fixed,
+    cases Fx : F (some x) with y,
+    { rwa [remove_none_none F Fx, F_none, option.some_inj, eq_comm] at x_fixed },
+    { exfalso, rw remove_none_some F ⟨y, Fx⟩ at x_fixed, exact F_derangement _ x_fixed } },
+  { intro h_opfp,
+    use equiv.perm.decompose_option.symm (some a, f),
+    split,
+    { intro x,
+      apply_fun (swap none (some a)),
+      simp only [perm.decompose_option_symm_apply, swap_apply_self, perm.coe_mul],
+      cases x,
+      { simp },
+      { simp only [equiv_functor.map_equiv_apply, equiv_functor.map,
+                    option.map_eq_map, option.map_some'],
+        by_cases x_vs_a : x = a,
+        { rw [x_vs_a, swap_apply_right], apply option.some_ne_none },
+        { have ne_1 : some x ≠ none := option.some_ne_none _,
+          have ne_2 : some x ≠ some a := (option.some_injective α).ne_iff.mpr x_vs_a,
+          rw [swap_apply_of_ne_of_ne ne_1 ne_2, (option.some_injective α).ne_iff],
+          intro contra,
+          exact x_vs_a (h_opfp contra) } } },
```

Comment:
```suggestion
      simp only [equiv_functor.map_equiv_apply, equiv_functor.map,
                  option.map_eq_map, option.map_some'],
      by_cases x_vs_a : x = a,
      { rw [x_vs_a, swap_apply_right], apply option.some_ne_none },
      have ne_1 : some x ≠ none := option.some_ne_none _,
      have ne_2 : some x ≠ some a := (option.some_injective α).ne_iff.mpr x_vs_a,
      rw [swap_apply_of_ne_of_ne ne_1 ne_2, (option.some_injective α).ne_iff],
      intro contra,
      exact x_vs_a (h_opfp contra) },
```
You don't need that much indenting

Reply:
if i've got two cases and the first is simple, i can just unindent the second then?